### PR TITLE
CL: add function to generate random `Positions` instances

### PR DESCRIPTION
**Pull request by pysel**
Closes: #4361

This function is used to generate an instance of the Positions struct with randomized field values.

The function populates `numSwaps` and `numAccounts` fields with a random number in the range of [2, 100]. For the remaining fields: `numFullRange`, `numNarrowRange`, `numConsecutive`, and `numOverlapping`, it generates a random number in the range of [0, numAccounts].

# Documentation

Minor

### CODE REVIEW

**Review by czarcas7ic — APPROVED** [review]

File: x/concentrated-liquidity/spread_rewards_test.go [hunk 1]
```diff
@@ -1359,6 +1360,27 @@ func (s *KeeperTestSuite) TestUpdatePosValueToInitValuePlusGrowthOutside() {
 	}
 }
 
+func NewRandomizedPositions() Positions {
```
Comment (p0mvn) [inline x/concentrated-liquidity/spread_rewards_test.go:1363]: So the function is made but why is it unused?

File: x/concentrated-liquidity/spread_rewards_test.go [hunk 2]
```diff
@@ -1359,6 +1360,27 @@ func (s *KeeperTestSuite) TestUpdatePosValueToInitValuePlusGrowthOutside() {
...
+	rand.Seed(time.Now().UnixNano())
```
Comment (p0mvn) [inline x/concentrated-liquidity/spread_rewards_test.go:1364]: Let's make the seed deterministic please

**Review by p0mvn — CHANGES_REQUESTED** [review]

File: x/concentrated-liquidity/spread_rewards_test.go (hunk 1 above)
Comment (pysel) [reply x/concentrated-liquidity/spread_rewards_test.go:1363]: I can change the concrete `Positions` struct initializations to use this function, the issue description only mentioned creating this function, hence, I left tests' code unchanged

File: x/concentrated-liquidity/spread_rewards_test.go (hunk 2 above)
Comment (pysel) [reply x/concentrated-liquidity/spread_rewards_test.go:1364]: but if we use a deterministic seed, won't it mean, that for each test, which gets a positions struct with this function, the initialized object will be the same? afaiu, there will be no point in using randomness at all

File: x/concentrated-liquidity/spread_rewards_test.go (hunk 1 above)
Comment (p0mvn) [reply x/concentrated-liquidity/spread_rewards_test.go:1363]: Can we please either make use of this or close this? I don't see the need for introducing unused code. The author probably implied making use of this

File: x/concentrated-liquidity/spread_rewards_test.go (hunk 2 above)
Comment (p0mvn) [reply x/concentrated-liquidity/spread_rewards_test.go:1364]: Can we find a way to use a different but deterministic and known seed for every use of `NewRandomizedPositions`?

Having a random seed based on time that developers don't know isn't very useful because even if you get a test failure, you can't reproduce it

File: x/concentrated-liquidity/spread_rewards_test.go (hunk 1 above)
Comment (pysel) [reply x/concentrated-liquidity/spread_rewards_test.go:1363]: makes sense

File: x/concentrated-liquidity/spread_rewards_test.go [hunk 3]
```diff
@@ -1359,6 +1360,27 @@ func (s *KeeperTestSuite) TestUpdatePosValueToInitValuePlusGrowthOutside() {
...
+
+	numSwaps := 2 + rand.Intn(99) // Random number between 2 and 100
+	numAccounts := 2 + rand.Intn(99)
+
+	numFullRange := rand.Intn(numAccounts + 1) // Random number between 0 and numAccounts inclusive
```
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/spread_rewards_test.go:1369]: It might be helpful to reference how we bound random inputs in the simulator. e.g. in this specific case, we probably want to have _some_ cap on the number of positions so tests don't run for 10+ minutes

Tldr is you can mod the output to bound the range of the final number, similar to what's done in #5476

**Comment by czarcas7ic** [comment]
@pysel bumping work on this, please lmk if you intend to address issues.

**Comment by pysel** [comment]
hey @czarcas7ic @AlpinYukseloglu. I made the seed deterministic and added a parameter to cap the max value of any field in randomized struct.

however, after trying to use the function in tests, it failed in one of them ([here](https://github.com/osmosis-labs/osmosis/pull/5382/files#diff-1a552fcc0a3d36556402345e840b658a7daccb101d7f73bca9ee2e3057ade2e6R2645)).

**Comment by czarcas7ic** [comment]
@pysel this is because the current tick liquidity has changed based off the num positions:

That seed should be calculated with a liquidity of `liq = Decimal("7801543753.113516346960781435")`

**Comment by pysel** [comment]
sure, I will finish that

**Comment by ValarDragon** [comment]
Closing this, thank you for the work. We've moved onto other methods of randomly testing this part of the code, and should phase out of this. Thank you for the time put in!
